sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/aggregator/DoubleHistogramAggregator.java

```diff
@@ -0,0 +1,158 @@
+  private final double[] boundaries;
+
+  // a cache for converting to MetricData
+  private final List<Double> boundaryList;
+
+  DoubleHistogramAggregator(
+      Resource resource,
+      InstrumentationLibraryInfo instrumentationLibraryInfo,
+      InstrumentDescriptor instrumentDescriptor,
+      double[] boundaries,
+      boolean stateful) {
+    super(resource, instrumentationLibraryInfo, instrumentDescriptor, stateful);
+    this.boundaries = boundaries;
+
+    List<Double> boundaryList = new ArrayList<>(this.boundaries.length);
+    for (double v : this.boundaries) {
+      boundaryList.add(v);
+    }
+    this.boundaryList = Collections.unmodifiableList(boundaryList);
+  }
+
+  @Override
+  public AggregatorHandle<HistogramAccumulation> createHandle() {
+    return new Handle(this.boundaries);
+  }
+
+  /**
+   * Return the result of the merge of two histogram accumulations. As long as one Aggregator
+   * instance produces all Accumulations with constant boundaries we don't need to worry about
+   * merging accumulations with different boundaries.
+   */
+  @Override
+  public final HistogramAccumulation merge(HistogramAccumulation x, HistogramAccumulation y) {
+    long[] mergedCounts = new long[x.getCounts().length];
+    for (int i = 0; i < x.getCounts().length; ++i) {
+      mergedCounts[i] = x.getCounts()[i] + y.getCounts()[i];
+    }
+    return HistogramAccumulation.create(x.getSum() + y.getSum(), mergedCounts);
+  }
+
+  @Override
+  public final MetricData toMetricData(
+      Map<Labels, HistogramAccumulation> accumulationByLabels,
+      long startEpochNanos,
+      long lastCollectionEpoch,
+      long epochNanos) {
+    return MetricData.createDoubleHistogram(
+        getResource(),
+        getInstrumentationLibraryInfo(),
+        getInstrumentDescriptor().getName(),
+        getInstrumentDescriptor().getDescription(),
+        getInstrumentDescriptor().getUnit(),
+        DoubleHistogramData.create(
+            isStateful() ? AggregationTemporality.CUMULATIVE : AggregationTemporality.DELTA,
+            MetricDataUtils.toDoubleHistogramPointList(
+                accumulationByLabels,
+                isStateful() ? startEpochNanos : lastCollectionEpoch,
+                epochNanos,
+                boundaryList)));
+  }
+
+  @Override
+  public HistogramAccumulation accumulateDouble(double value) {
+    long[] counts = new long[this.boundaries.length + 1];
+    counts[findBucketIndex(this.boundaries, value)] = 1;
+    return HistogramAccumulation.create(value, counts);
+  }
+
+  @Override
+  public HistogramAccumulation accumulateLong(long value) {
+    return accumulateDouble((double) value);
+  }
+
+  // Benchmark shows that linear search performs better than binary search with ordinary
+  // buckets.
+  private static int findBucketIndex(double[] boundaries, double value) {
+    for (int i = 0; i < boundaries.length; ++i) {
+      if (Double.compare(value, boundaries[i]) <= 0) {
```

Comment:
```suggestion
      if (value <= boundaries[i]) {
```

Do we have to handle NAN? If so, I'd handle it specially, it's not clear to me what it means in the case of a histogram like this. I suspect we'd want the `MeterProvider` to filter out NANs and we use simple comparisons here though.

Reply:
I don't see any contracts about `NaN` in the specifications so just being defensive here.